### ISSUE TITLE
Report the position of a Workload in a queue via the visibility API

### ISSUE THREAD

**Issue by alculquicondor** [kubernetes-sigs/kueue#1657]
**What would you like to be added**:

Provide a visibility endpoint that returns the position of a workload in a LocalQueue or ClusterQueue.

This is a continuation of https://github.com/kubernetes-sigs/kueue/issues/168

**Why is this needed**:

To improve visibility to end users so they can estimate how long they will have to wait for their workloads to run

**Completion requirements**:

This enhancement requires the following artifacts:

- [x] Design doc
- [x] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

**Comment by alculquicondor** [comment]
Note that this issue requires a design (in the keps/ folder) prior to implementation.

**Comment by alculquicondor** [comment]
The best starting point would be to read the existing KEP for the visibility API.
https://github.com/kubernetes-sigs/kueue/tree/main/keps/168-2-pending-workloads-visibility

**Comment by highpon** [comment]
@alculquicondor 
I have a question.
Is what you want to accomplish with this issue to estimate the number of minutes until the pending workload is executed?

**Comment by tenzen-y** [comment]
> @alculquicondor I have a question. Is what you want to accomplish with this issue to estimate the number of minutes until the pending workload is executed?

From my understanding, the answer is no. I would like to assume the goal of this enhancement is to provide the ability to specify workload name query parameters something like `/apis/visibility.kueue.x-k8s.io/v1alpha1/namespaces/default/localqueues/user-queue/pendingworkloads?workload=myworkload`.

**Comment by alculquicondor** [comment]
yeah, we only want to know what is the current position of the workload in the queue.
We don't have the means to estimate the wait time at the moment (Kueue doesn't have long term memory).

However, this feature is a step towards that.

**Comment by alaypatel07** [comment]
If this has not been worked on, I can help with making this feature possible. @alculquicondor @tenzen-y

**Comment by highpon** [comment]
@alaypatel07 
I have been too busy with work lately to work on this issue.
If you are able to get to it, I would be very grateful for your help in resolving this issue!

**Comment by tenzen-y** [comment]
@alaypatel07 If you want to work on this issue, feel free to assign yourself with `/assign`.

**Comment by alaypatel07** [comment]
yes I can help with this issue, thanks folks.

/assign

**Comment by alaypatel07** [comment]
@tenzen-y @alculquicondor I have been playing around the visibility API and I can create the kep for implementing
```
/apis/visibility.kueue.x-k8s.io/v1alpha1/namespaces/default/localqueues/user-queue/pendingworkloads?workload=myworkload
```
Since workload is a namespaced resource, I wonder if this feature should only apply on localqueues?

The issue with fetching the workload from cluster queue is two fold:
1. in order to correctly resolve the workload name in cluster queue, the URI will require additional arg workloadNamespace, something like this:
```
kubectl get --raw "/apis/visibility.kueue.x-k8s.io/v1alpha1/clusterqueues/cluster-queue/pendingworkloads?workloadName=myworkload&workloadNamespace=default"
```
2. The information on where the workload is in global queue is available from localqueue response in `positionInClusterQueue` field. 

Any thoughts on this side will be helpful, TIA

**Comment by alculquicondor** [comment]
I don't think you should be piggy-backing on the existing URLs.
We should have a new URL just for workloads, something like this:

`/apis/visibility.kueue.x-k8s.io/v1alpha1/namespaces/<ns>/workloads/<name>`

The next question is whether it should present any information if it's not pending, and what it should present.

**Comment by alaypatel07** [comment]
@alculquicondor out of curiosity why would having a new URL be preferable?

IIUC, visibility API w.r.t. workload only provides information about a workload on the queue and extending the existing URL will give users a way to filter the 1 WL out of all the pending ones
 May be in the future have an additional arg of label selector as query to filter from pending workloads URL.

Are there any use case that I'm missing which will be achieved better with a separate URL?

**Comment by alculquicondor** [comment]
Think of it the other way: I want a way to know all the queueing information for a particular workload.

Then the endpoint gives me in which queue it's queued and what its position.
